File: apps/health/internal/projection/gaps.go
package projection

import (
	"database/sql"
	"fmt"
	"sort"
	"time"
)

// GapReport is the local calendar days between first and last data
// that have no observation (or episode) start in that day.
type GapReport struct {
	Kind    string   `json:"kind,omitempty"` // empty = overall
	Table   string   `json:"table,omitempty"`
	First   string   `json:"first,omitempty"` // YYYY-MM-DD local
	Last    string   `json:"last,omitempty"`
	Present []string `json:"present,omitempty"`
	Missing []string `json:"missing,omitempty"`
}

// PresentN is how many local days have at least one row.
func (g GapReport) PresentN() int { return len(g.Present) }

// MissingN is how many local days in [First, Last] have no row.
func (g GapReport) MissingN() int { return len(g.Missing) }

// SpanDays is inclusive calendar days from first to last, or 0 if empty.
func (g GapReport) SpanDays() int {
	if g.First == "" || g.Last == "" {
		return 0
	}
	a, err1 := time.Parse("2006-01-02", g.First)
	b, err2 := time.Parse("2006-01-02", g.Last)
	if err1 != nil || err2 != nil {
		return 0
	}
	return int(b.Sub(a).Hours()/24) + 1
}

// ObservationGaps lists missing local days for one observation kind
// between its first and last sample. Empty kind means every observation.
func ObservationGaps(db *sql.DB, kind string, loc *time.Location) (GapReport, error) {
	if loc == nil {
		loc = time.UTC
	}
	q := `SELECT start_ts FROM observations`
	var args []any
	if kind != "" {
		q += ` WHERE kind = ?`
		args = append(args, kind)
	}
	q += ` ORDER BY start_ts`
	return scanGaps(db, q, args, kind, "observations", loc)
}

// EpisodeGaps is the same scan over episodes.
func EpisodeGaps(db *sql.DB, kind string, loc *time.Location) (GapReport, error) {
	if loc == nil {
		loc = time.UTC
	}
	q := `SELECT start_ts FROM episodes`
	var args []any
	if kind != "" {
		q += ` WHERE kind = ?`
		args = append(args, kind)
	}
	q += ` ORDER BY start_ts`
	return scanGaps(db, q, args, kind, "episodes", loc)
}

// OverallGaps is days with no observation of any kind, plus no episode.
func OverallGaps(db *sql.DB, loc *time.Location) (GapReport, error) {
	if loc == nil {
		loc = time.UTC
	}
	obs, err := ObservationGaps(db, "", loc)
	if err != nil {
		return GapReport{}, err
	}
	eps, err := EpisodeGaps(db, "", loc)
	if err != nil {
		return GapReport{}, err
	}
	set := map[string]bool{}
	for _, d := range obs.Present {
		set[d] = true
	}
	for _, d := range eps.Present {
		set[d] = true
	}
	if len(set) == 0 {
		return GapReport{Table: "overall"}, nil
	}
	present := make([]string, 0, len(set))
	for d := range set {
		present = append(present, d)
	}
	sort.Strings(present)
	g := GapReport{
		Table:   "overall",
		First:   present[0],
		Last:    present[len(present)-1],
		Present: present,
		Missing: missingBetween(present[0], present[len(present)-1], set),
	}
	return g, nil
}

func scanGaps(db *sql.DB, q string, args []any, kind, table string, loc *time.Location) (GapReport, error) {
	rows, err := db.Query(q, args...)
	if err != nil {
		return GapReport{}, err
	}
	defer rows.Close()
	set := map[string]bool{}
	for rows.Next() {
		var ts string
		if err := rows.Scan(&ts); err != nil {
			return GapReport{}, err
		}
		day, err := localDay(ts, loc)
		if err != nil {
			return GapReport{}, err
		}
		set[day] = true
	}
	if err := rows.Err(); err != nil {
		return GapReport{}, err
	}
	g := GapReport{Kind: kind, Table: table}
	if len(set) == 0 {
		return g, nil
	}
	present := make([]string, 0, len(set))
	for d := range set {
		present = append(present, d)
	}
	sort.Strings(present)
	g.First = present[0]
	g.Last = present[len(present)-1]
	g.Present = present
	g.Missing = missingBetween(g.First, g.Last, set)
	return g, nil
}

func localDay(ts string, loc *time.Location) (string, error) {
	t, err := time.Parse(time.RFC3339Nano, ts)
	if err != nil {
		return "", fmt.Errorf("start_ts %q: %w", ts, err)
	}
	return t.In(loc).Format("2006-01-02"), nil
}

func missingBetween(first, last string, present map[string]bool) []string {
	a, err1 := time.Parse("2006-01-02", first)
	b, err2 := time.Parse("2006-01-02", last)
	if err1 != nil || err2 != nil {
		return nil
	}
	var out []string
	for d := a; !d.After(b); d = d.AddDate(0, 0, 1) {
		s := d.Format("2006-01-02")
		if !present[s] {
			out = append(out, s)
		}
	}
	return out
}

// LatestLocalDay is the most recent local calendar day that has any observation.
func LatestLocalDay(db *sql.DB, loc *time.Location) (string, error) {
	if loc == nil {
		loc = time.UTC
	}
	var ts sql.NullString
	if err := db.QueryRow(`SELECT MAX(start_ts) FROM observations`).Scan(&ts); err != nil {
		return "", err
	}
	if !ts.Valid || ts.String == "" {
		return "", nil
	}
	return localDay(ts.String, loc)
}

File: core/health-core/src/gaps.rs
//! Local-calendar gap reports. "None seen" is not completeness (ADR 0008 R8).

use std::collections::BTreeSet;

use rusqlite::Connection;
use serde::Serialize;

use crate::{Error, Result};

#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize)]
pub struct GapReport {
    #[serde(skip_serializing_if = "String::is_empty")]
    pub kind: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub table: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub first: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub last: String,
    pub present: Vec<String>,
    pub missing: Vec<String>,
}

impl GapReport {
    pub fn present_n(&self) -> usize {
        self.present.len()
    }

    pub fn missing_n(&self) -> usize {
        self.missing.len()
    }

    pub fn span_days(&self) -> i64 {
        match (parse_day(&self.first), parse_day(&self.last)) {
            (Some(a), Some(b)) if b >= a => (b - a) + 1,
            _ => 0,
        }
    }
}

/// Missing local days for one observation kind, or every observation.
pub fn observation_gaps(db: &Connection, kind: Option<&str>) -> Result<GapReport> {
    scan_gaps(db, "observations", kind)
}

/// Missing local days for one episode kind, or every episode.
pub fn episode_gaps(db: &Connection, kind: Option<&str>) -> Result<GapReport> {
    scan_gaps(db, "episodes", kind)
}

/// Days with neither an observation nor an episode.
pub fn overall_gaps(db: &Connection) -> Result<GapReport> {
    let obs = observation_gaps(db, None)?;
    let eps = episode_gaps(db, None)?;
    let mut set = BTreeSet::new();
    set.extend(obs.present);
    set.extend(eps.present);
    if set.is_empty() {
        return Ok(GapReport {
            table: "overall".into(),
            ..GapReport::default()
        });
    }
    let present: Vec<String> = set.iter().cloned().collect();
    let first = present[0].clone();
    let last = present[present.len() - 1].clone();
    let missing = missing_between(&first, &last, &set);
    Ok(GapReport {
        table: "overall".into(),
        first,
        last,
        present,
        missing,
        ..GapReport::default()
    })
}

fn scan_gaps(db: &Connection, table: &str, kind: Option<&str>) -> Result<GapReport> {
    let mut sql = format!("SELECT start_ts FROM {table}");
    if kind.is_some() {
        sql.push_str(" WHERE kind = ?1");
    }
    sql.push_str(" ORDER BY start_ts");
    let mut stmt = db.prepare(&sql)?;
    let mut set = BTreeSet::new();
    if let Some(kind) = kind {
        let rows = stmt.query_map([kind], |row| row.get::<_, String>(0))?;
        for ts in rows {
            set.insert(local_day(&ts?)?);
        }
    } else {
        let rows = stmt.query_map([], |row| row.get::<_, String>(0))?;
        for ts in rows {
            set.insert(local_day(&ts?)?);
        }
    }
    let mut g = GapReport {
        kind: kind.unwrap_or("").to_string(),
        table: table.to_string(),
        ..GapReport::default()
    };
    if set.is_empty() {
        return Ok(g);
    }
    g.present = set.iter().cloned().collect();
    g.first = g.present[0].clone();
    g.last = g.present[g.present.len() - 1].clone();
    g.missing = missing_between(&g.first, &g.last, &set);
    Ok(g)
}

fn local_day(ts: &str) -> Result<String> {
    if ts.len() >= 10 {
        Ok(ts[..10].to_string())
    } else {
        Err(Error::Invalid(format!("start_ts {ts:?}")))
    }
}

fn parse_day(day: &str) -> Option<i64> {
    if day.len() != 10 {
        return None;
    }
    let y: i64 = day[0..4].parse().ok()?;
    let m: i64 = day[5..7].parse().ok()?;
    let d: i64 = day[8..10].parse().ok()?;
    Some(y * 400 + m * 32 + d)
}

fn missing_between(first: &str, last: &str, present: &BTreeSet<String>) -> Vec<String> {
    let mut out = Vec::new();
    let mut cur = first.to_string();
    while cur.as_str() <= last {
        if !present.contains(&cur) {
            out.push(cur.clone());
        }
        if cur == last {
            break;
        }
        cur = next_day(&cur);
        if cur.is_empty() {
            break;
        }
    }
    out
}

fn next_day(day: &str) -> String {
    let Ok(y) = day[0..4].parse::<i32>() else {
        return String::new();
    };
    let Ok(m) = day[5..7].parse::<u32>() else {
        return String::new();
    };
    let Ok(d) = day[8..10].parse::<u32>() else {
        return String::new();
    };
    let days = days_in_month(y, m);
    if d < days {
        return format!("{y:04}-{m:02}-{:02}", d + 1);
    }
    if m < 12 {
        return format!("{y:04}-{:02}-01", m + 1);
    }
    format!("{:04}-01-01", y + 1)
}

fn days_in_month(year: i32, month: u32) -> u32 {
    match month {
        1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
        4 | 6 | 9 | 11 => 30,
        2 if is_leap(year) => 29,
        2 => 28,
        _ => 31,
    }
}

fn is_leap(year: i32) -> bool {
    (year % 4 == 0 && year % 100 != 0) || year % 400 == 0
}
